## Problem 15: Lattice Paths

Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

How many such routes are there through a 20×20 grid?

In [1]:
# Need to make 20 right moves and 20 down moves, 40 moves in total.
# One right move is indistinguishable from another: Rrd is the same as rRd.

# First move is either a right or down: (20/40 + 20/40)
# Second move is either a right or down: (20/40 * 19/39 + 20/40 * 20/39 + 20/40 * 19/39 + 20/40 * 20/39)

# 20 * 19 * 18 * 17 * ...
# -----------------------
# 40 * 39 * 38 * 37 * ...

# Number of combinations: 40 C 20

In [2]:
''' 
Calculates the factorial of a number using recursion. Cache the values so we only need to caluclate (2 * grid_size)!, then
the other factorials in nCr will have been cached.

Base case : 1 : 1
Other case : n : n *= factorial(n-1)

eg:
>>> factorial(5)
>>> 120
'''

from functools import lru_cache

@lru_cache(maxsize=1_000)
def factorial(number : int) -> int:
    
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Calculates the factorial of a number using recursion. Cache the values so we only need to caluclate (2 * grid_size)!, 
        then the other factorials in nCr will have been cached.

        Base case : 1 : 1
        Other case : n : n *= factorial(n-1)
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number to calculate the factorial of.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : The factorial of the number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> factorial(5)
        >>> 120
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 1: raise ValueError('Please enter an integer > 0 for the number argument.')
    
    if number == 1:
        return 1
    else:
        return number * factorial(number - 1)

In [3]:
''' 
Calculates the binomail function nCr.

nCr = n! / (r! (n-r)!)

eg:
>>> nCr(4,2)
>>> 6
'''

def nCr(n : int, r : int) -> int:
    
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Calculates the binomail function nCr.

        nCr = n! / (r! (n-r)!)
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : n : The total number to choose from.
        : int : n : The number to choose.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : nCr of n and r: nCr = n! / (r! (n-r)!)
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> nCr(4,2)
        >>> 6
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(n, int): raise ValueError('Please enter an integer > 0 for the n argument.')
    if n < 1: raise ValueError('Please enter an integer > 0 for the n argument.')
    if not isinstance(r, int): raise ValueError('Please enter an integer > 0 for the r argument.')
    if r < 1: raise ValueError('Please enter an integer > 0 for the r argument.')
    
    return int(factorial(n) / ((factorial(r)) * (factorial(n - r))))

In [4]:
grid_size = 20
solution = nCr(2 * grid_size, grid_size)

print(f'The number of routes are there through a {grid_size}×{grid_size} grid, starting at the top left corner and only being able to move to the right') 
print(f'and down, is {solution:,}.')

The number of routes are there through a 20×20 grid, starting at the top left corner and only being able to move to the right
and down, is 137,846,528,820.


### Problem 15 Solution:

The number of routes are there through a 20×20 grid, starting at the top left corner and only being able to move to the right and down, is 137,846,528,820.